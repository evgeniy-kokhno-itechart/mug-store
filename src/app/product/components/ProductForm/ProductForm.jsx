import React from 'react';
import * as Yup from 'yup';
import _ from 'lodash';
import { PropTypes } from 'prop-types';
import {
  FormBase, Input, TextArea, Dropdown,
} from '../../../shared';
import './ProductForm.scss';


class ProductForm extends FormBase {
  state = {
    data: {
      id: '',
      title: '',
      imageURL: '',
      description: '',
      categoryId: '',
      discount: 0,
      producer: '',
      currentCurrencyPrice: 0,
      rate: 0,
    },
    errors: {},
  };

  objectSchema = {
    id: Yup.string(),
    imageURL: Yup.string()
      .min(5)
      .label('Image URL'),
    title: Yup.string()
      .required()
      .min(3)
      .label('Title'),
    description: Yup.string()
      .required()
      .max(500)
      .label('Details'),
    categoryId: Yup.number()
      .typeError('Please choose category')
      .required()
      .label('Category'),
    currentCurrencyPrice: Yup.number()
      .typeError('Please enter a valid number')
      .required()
      .moreThan(0)
      .label('Price'),
    discount: Yup.number()
      .typeError('Please enter a valid number')
      .min(0)
      .max(100)
      .label('Discount, %'),
    producer: Yup.string()
      .required()
      .min(5)
      .label('Producer'),
    publishDate: Yup.date(),
    rate: Yup.number()
      .typeError('Please enter a valid number')
      .required()
      .min(0)
      .max(5)
      .label('Rate'),
  };

  componentDidMount() {
    if (this.props.product.id) {
      this.setState({ data: this.props.product });
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { product } = this.props;

    if (prevState.data.id !== product.id) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({ data: product });
    }
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const errors = this.validateForm();
    this.setState({ errors: errors || {} });
    if (errors) {
      return;
    }
    this.props.onSubmit(this.state.data);
  };

  render() {
    const { data, errors } = this.state;
    const { categories, currentCurrencyName } = this.props;

    return (
      <div className="product-form">
        <h1 className="product-form__header">Product Info</h1>
        <form className="product-form__content prod-form-content" onSubmit={this.handleSubmit}>
          <Input
            type="text"
            name="title"
            label="Title"
            value={data.title}
            error={errors.title}
            validationSchema={this.objectSchema.title}
            onChange={this.handleControlChange}
          />

          <Input
            type="text"
            name="imageURL"
            label="Image URL"
            value={_.get(data, 'imageURL')}
            error={errors.imageURL}
            validationSchema={this.objectSchema.imageURL}
            onChange={this.handleControlChange}
          />

          <TextArea
            name="description"
            label="Details"
            value={data.description}
            error={errors.description}
            validationSchema={this.objectSchema.description}
            onChange={this.handleControlChange}
          />

          <Dropdown
            name="categoryId"
            label="Category"
            options={categories}
            value={data.categoryId}
            error={errors.categoryId}
            defaultText="Please choose..."
            wrapperClasses="category-dropdown"
            labelClasses="category-dropdown__label"
            selectClasses="category-dropdown__selector"
            validationSchema={this.objectSchema.categoryId}
            onChange={this.handleControlChange}
          />

          <Input
            type="text"
            name="currentCurrencyPrice"
            label={`Price, ${currentCurrencyName}`}
            value={_.get(data, 'currentCurrencyPrice')}
            error={errors.currentCurrencyPrice}
            validationSchema={this.objectSchema.currentCurrencyPrice}
            onChange={this.handleControlChange}
          />

          <Input
            type="text"
            name="discount"
            label="Dicount, %"
            value={_.get(data, 'discount')}
            error={errors.discount}
            validationSchema={this.objectSchema.discount}
            onChange={this.handleControlChange}
          />

          <Input
            type="text"
            name="producer"
            label="Producer"
            value={_.get(data, 'producer')}
            error={errors.producer}
            validationSchema={this.objectSchema.producer}
            onChange={this.handleControlChange}
          />

          <Input
            type="text"
            name="rate"
            label="Rate"
            value={_.get(data, 'rate')}
            error={errors.rate}
            validationSchema={this.objectSchema.rate}
            onChange={this.handleControlChange}
          />

          <button
            type="submit"
            disabled={this.validateForm(this.objectSchema, data)}
            className="button button--solid prod-form-button"
          >
            Save
          </button>
        </form>
      </div>
    );
  }
}

ProductForm.propTypes = {
  currentCurrencyName: PropTypes.string.isRequired,

  categories: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string, name: PropTypes.string })).isRequired,

  product: PropTypes.shape({
    id: PropTypes.string,
    title: PropTypes.string,
    imageURL: PropTypes.string,
    categoryId: PropTypes.string,
    description: PropTypes.string,
    producer: PropTypes.string,
    discount: PropTypes.number,
    currentCurrencyPrice: PropTypes.number,
    rate: PropTypes.string,
  }),

  onSubmit: PropTypes.func.isRequired,
};

ProductForm.defaultProps = {
  product: {
    id: '',
    imageURL: '',
    title: 'defaultProduct',
    description: '',
    categoryId: '',
    currentCurrencyPrice: 0,
    discount: 0,
    producer: '',
    rate: '',
  },
};

export default ProductForm;
