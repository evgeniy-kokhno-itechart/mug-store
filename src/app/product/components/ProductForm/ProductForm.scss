.product-form {
    &__header {
        text-align: center;
        margin: 1.5rem;
    }

    &__content {
        width: 50%;
        margin: 0 auto;
    }
}

.category-dropdown {
    &__label {
        margin-bottom: 0.5rem;
    }

    &__selector {
        width: 100%;
        padding: 0.4rem 0.5rem;
    }
}

.prod-form-content {
    &__button {
        margin: 0.5rem 0;
        width: 100%;
    }
}