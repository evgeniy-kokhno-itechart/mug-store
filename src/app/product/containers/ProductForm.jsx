/* eslint-disable no-nested-ternary */
/* eslint-disable react/no-did-update-set-state */
import React, { Component } from 'react';
import * as Yup from 'yup';
import _ from 'lodash';
import { PropTypes } from 'prop-types';
import { connect } from 'react-redux';
import { getProduct, saveProduct, clearCurrentProductInfo } from '../ProductsActions';
import { productCostSelector } from '../ProductsSelectors';
import { getCategories } from '../../catalog';
import {
  FormService, Input, Spinner, ErrorMessage, TextArea, Dropdown,
} from '../../shared';


export class ProductForm extends Component {
  state = {
    data: {
      id: '',
      title: '',
      imageURL: '',
      description: '',
      categoryId: '',
      discount: 0,
      producer: '',
      currentCurrencyPrice: 0,
      rate: 0,
    },
    errors: {},
  };

  productObjectSchema = {
    id: Yup.string(),
    imageURL: Yup.string()
      .min(5)
      .label('Image URL'),
    title: Yup.string()
      .required()
      .min(3)
      .label('Title'),
    description: Yup.string()
      .required()
      .max(500)
      .label('Details'),
    categoryId: Yup.number()
      .typeError('Please choose category')
      .required()
      .label('Category'),
    currentCurrencyPrice: Yup.number()
      .typeError('Please enter a valid number')
      .required()
      .moreThan(0)
      .label('Price'),
    discount: Yup.number()
      .typeError('Please enter a valid number')
      .min(0)
      .max(100)
      .label('Discount, %'),
    producer: Yup.string()
      .required()
      .min(5)
      .label('Producer'),
    publishDate: Yup.date(),
    rate: Yup.number()
      .typeError('Please enter a valid number')
      .required()
      .min(0)
      .max(5)
      .label('Rate'),
  };

  componentDidMount() {
    const productId = this.props.match.params.id;
    if (productId === 'new') {
      return;
    }

    this.props.getProduct(productId);
  }

  componentDidUpdate(prevProps, prevState) {
    const { productState, currentCurrency } = this.props;

    if (currentCurrency !== prevProps.currentCurrency) {
      const changedData = { ...prevState.data };
      changedData.currentCurrencyPrice = productState.product.currentCurrencyPrice;
      this.setState({ data: changedData });
    }

    if (productState.product.id !== prevState.data.id) {
      this.setState({ data: this.mapToViewModel(this.props.productState.product) });
    }
  }

  componentWillUnmount() {
    this.props.clearCurrentProductInfo();
  }

  displayIsLoading = () => {
    const { productState } = this.props;
    if (productState.loadingStatus.isInProcess || productState.savingStatus.isInProcess) {
      return <Spinner spinnerClasses="spinner--medium" wrapperClasses="mt-5" />;
    }
    return null;
  }

  displayGotError = () => {
    const { productState } = this.props;
    if (productState.loadingStatus.hasFailed || productState.savingStatus.hasFailed) {
      return <ErrorMessage message={productState.loadingStatus.error || productState.savingStatus.error} />;
    }
    return null;
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const errors = FormService.validateForm(this.productObjectSchema, this.state.data);
    this.setState({ errors: errors || {} });
    if (errors) {
      return;
    }

    // eslint-disable-next-line no-shadow
    const { saveProduct } = this.props;
    const product = this.mapFromViewModel(this.state.data);
    saveProduct(product, '/catalog');
  };

  handleChange = (e, matchedInputName) => {
    const { target: input } = e;

    this.setState(prevState => FormService.handleChange(input, matchedInputName, prevState, this.productObjectSchema));
  };

  mapToViewModel(product) {
    const dataForReturn = {
      id: product.id,
      title: product.title,
      imageURL: product.imageURL,
      description: product.description,
      categoryId: product.category.id,
      currentCurrencyPrice: product.currentCurrencyPrice,
      discount: product.discount,
      producer: product.producer,
      publishDate: product.publishDate,
      rate: product.rate,
    };
    return dataForReturn;
  }

  mapFromViewModel(data) {
    return {
      id: data.id,
      title: data.title,
      imageURL: data.imageURL,
      description: data.description,
      category: this.props.categories.find(category => category.id === data.categoryId),
      basePrice: data.currentCurrencyPrice / this.props.currentCurrency.rate,
      discount: data.discount,
      producer: data.producer,
      rate: data.rate,
    };
  }

  renderForm = () => {
    const { data, errors } = this.state;
    const { categories, currentCurrency } = this.props;

    return (
      <React.Fragment>
        <h1 className="text-center m-3">Product Info</h1>
        <form className="col-10 col-md-8 col-lg-7 col-xl-5 mx-auto" onSubmit={this.handleSubmit}>
          <Input
            type="text"
            name="title"
            label="Title"
            value={_.get(data, 'title')}
            error={errors.title}
            onValueChange={this.handleChange}
          />

          <Input
            type="text"
            name="imageURL"
            label="Image URL"
            value={_.get(data, 'imageURL')}
            error={errors.imageURL}
            onValueChange={this.handleChange}
          />

          <TextArea
            name="description"
            label="Details"
            value={data.description}
            error={errors.description}
            onValueChange={this.handleChange}
          />

          <Dropdown
            name="categoryId"
            label="Category"
            options={categories}
            value={data.categoryId}
            error={errors.categoryId}
            defaultText="Please choose..."
            selectClasses="form-control"
            onChange={this.handleChange}
          />

          <Input
            type="text"
            name="currentCurrencyPrice"
            label={`Price, ${currentCurrency.name}`}
            value={_.get(data, 'currentCurrencyPrice')}
            error={errors.currentCurrencyPrice}
            onValueChange={this.handleChange}
          />

          <Input
            type="text"
            name="discount"
            label="Dicount, %"
            value={_.get(data, 'discount')}
            error={errors.discount}
            onValueChange={this.handleChange}
          />

          <Input
            type="text"
            name="producer"
            label="Producer"
            value={_.get(data, 'producer')}
            error={errors.producer}
            onValueChange={this.handleChange}
          />

          <Input type="text" name="rate" label="Rate" value={_.get(data, 'rate')} error={errors.rate} onValueChange={this.handleChange} />

          <button type="submit" disabled={FormService.validateForm(this.productObjectSchema, data)} className="btn btn-secondary w-100">
            Save
          </button>
        </form>
      </React.Fragment>
    );
  };

  render() {
    return this.displayIsLoading() || this.displayGotError() || this.renderForm();
  }
}

ProductForm.propTypes = {
  match: PropTypes.shape({ params: PropTypes.shape({ id: PropTypes.string }) }),

  currentCurrency: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
    rate: PropTypes.number,
  }).isRequired,

  categories: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string, name: PropTypes.string })),

  productState: PropTypes.shape({
    product: PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      imageURL: PropTypes.string.isRequired,
      category: PropTypes.shape({ id: PropTypes.string, name: PropTypes.string }),
      description: PropTypes.string,
      producer: PropTypes.string,
      discount: PropTypes.number,
      currentCurrencyPrice: PropTypes.number,
      rate: PropTypes.string,
    }),
    loadingStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
    savingStatus: PropTypes.shape({
      isInProcess: PropTypes.bool,
      hasFailed: PropTypes.bool,
      error: PropTypes.string,
    }),
  }),

  getProduct: PropTypes.func.isRequired,
  saveProduct: PropTypes.func.isRequired,
  clearCurrentProductInfo: PropTypes.func.isRequired,
};

ProductForm.defaultProps = {
  categories: [],
  match: { params: { id: '' } },

  productState: {
    product: {
      id: '',
      imageURL: '',
      title: 'defaultProduct',
      description: '',
      category: {},
      currentCurrencyPrice: 0,
      discount: 0,
      producer: '',
      rate: '',
    },
    loadingStatus: {
      isInProcess: true,
      hasFailed: false,
      error: '',
    },
    savingStatus: {
      isInProcess: true,
      hasFailed: false,
      error: '',
    },
  },
};

const mapStateToProps = state => ({
  currentCurrency: state.currency.currentCurrency,
  categories: state.category.categories,

  productState: {
    product: productCostSelector(state),
    loadingStatus: state.products.currentProductStatus,
    savingStatus: state.products.savingStatus,
  },
});

const mapDispatchToProps = {
  getCategories,
  getProduct,
  saveProduct,
  clearCurrentProductInfo,
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ProductForm);
