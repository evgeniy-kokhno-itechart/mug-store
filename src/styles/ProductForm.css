.product-form__spinner {
  width: 3rem;
  height: 3rem;
}
