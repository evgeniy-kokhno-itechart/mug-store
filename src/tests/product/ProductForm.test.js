/* eslint-disable no-undef */
import React from 'react';

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import ProductForm from '../../app/product/components/ProductForm/ProductForm';

configure({ adapter: new Adapter() });

describe('<ProductForm />', () => {
  let ProductFormWrapper;
  let submitForm;

  beforeEach(() => {
    submitForm = jest.fn();
  });

  it('renders empty form with default props', () => {
    ProductFormWrapper = shallow(
      <ProductForm
        currentCurrencyName='CURR1'
        categories={[{ id: '1', name: 'category1' }, { id: '2', name: 'category2' }]}
        onSubmit={submitForm}
      />,
    );
    expect(ProductFormWrapper).toMatchSnapshot();
  });

  it('renders not empty fields with product data provided', () => {
    ProductFormWrapper = shallow(
      <ProductForm
        currentCurrencyName='CURR1'
        categories={[{ id: '1', name: 'category1' }, { id: '2', name: 'category2' }]}
        onSubmit={submitForm}
        product={{
          id: '1',
          imageURL: '#',
          title: 'Mug 1',
          description: 'product 1 description',
          categoryId: '2',
          basePrice: 5,
          currentCurrencyPrice: 5,
          discount: 0,
          producer: 'test producer 1',
          publishDate: 'Tue Aug 20 2019 16:27:02 GMT+0300 (Moscow Standard Time)',
          rate: '5',
        }}
      />,
    );
    expect(ProductFormWrapper).toMatchSnapshot();
  });

  it('submits form properly after the from has been filled out', () => {
    ProductFormWrapper = shallow(
      <ProductForm
        currentCurrencyName='CURR1'
        categories={[{ id: '1', name: 'category1' }, { id: '2', name: 'category2' }]}
        onSubmit={submitForm}
      />,
    );
    ProductFormWrapper.setState({
      data: {
        imageURL: '#123456',
        title: 'Test Mug',
        description: 'product 1 description',
        categoryId: '1',
        currentCurrencyPrice: 5,
        discount: 0,
        producer: 'test producer 1',
        rate: '5',
      },
    });
    const form = ProductFormWrapper.find('form');
    form.simulate('submit', { preventDefault() {} });

    expect(submitForm).toHaveBeenCalled();
  });
});
