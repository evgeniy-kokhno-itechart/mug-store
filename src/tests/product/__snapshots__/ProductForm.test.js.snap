// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductForm /> renders empty form with default props 1`] = `
<Fragment>
  <h1
    className="text-center m-3"
  >
    Product Info
  </h1>
  <form
    className="col-10 col-md-8 col-lg-7 col-xl-5 mx-auto"
    onSubmit={[Function]}
  >
    <Input
      error=""
      label="Title"
      name="title"
      onChange={[Function]}
      type="text"
      validationSchema={
        StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "min": true,
            "required": true,
          },
          "_label": "Title",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value=""
    />
    <Input
      error=""
      label="Image URL"
      name="imageURL"
      onChange={[Function]}
      type="text"
      validationSchema={
        StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "min": true,
          },
          "_label": "Image URL",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value=""
    />
    <TextArea
      error=""
      label="Details"
      name="description"
      onChange={[Function]}
      validationSchema={
        StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "max": true,
            "required": true,
          },
          "_label": "Details",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value=""
    />
    <Dropdown
      defaultText="Please choose..."
      error=""
      label="Category"
      labelClasses=""
      name="categoryId"
      onChange={[Function]}
      options={
        Array [
          Object {
            "id": "1",
            "name": "category1",
          },
          Object {
            "id": "2",
            "name": "category2",
          },
        ]
      }
      selectClasses="form-control"
      validationSchema={
        NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_label": "Category",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value=""
      wrapperClasses=""
    />
    <Input
      error=""
      label="Price, CURR1"
      name="currentCurrencyPrice"
      onChange={[Function]}
      type="text"
      validationSchema={
        NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "min": true,
            "required": true,
          },
          "_label": "Price",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value={0}
    />
    <Input
      error=""
      label="Dicount, %"
      name="discount"
      onChange={[Function]}
      type="text"
      validationSchema={
        NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "max": true,
            "min": true,
          },
          "_label": "Discount, %",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value={0}
    />
    <Input
      error=""
      label="Producer"
      name="producer"
      onChange={[Function]}
      type="text"
      validationSchema={
        StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "min": true,
            "required": true,
          },
          "_label": "Producer",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value=""
    />
    <Input
      error=""
      label="Rate"
      name="rate"
      onChange={[Function]}
      type="text"
      validationSchema={
        NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "max": true,
            "min": true,
            "required": true,
          },
          "_label": "Rate",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value={0}
    />
    <button
      className="btn btn-secondary w-100"
      disabled={
        Object {
          "categoryId": "Please choose category",
          "currentCurrencyPrice": "Price must be greater than 0",
          "description": "Details is a required field",
          "imageURL": "Image URL must be at least 5 characters",
          "producer": "Producer must be at least 5 characters",
          "title": "Title must be at least 3 characters",
        }
      }
      type="submit"
    >
      Save
    </button>
  </form>
</Fragment>
`;

exports[`<ProductForm /> renders not empty fields with product data provided 1`] = `
<Fragment>
  <h1
    className="text-center m-3"
  >
    Product Info
  </h1>
  <form
    className="col-10 col-md-8 col-lg-7 col-xl-5 mx-auto"
    onSubmit={[Function]}
  >
    <Input
      error=""
      label="Title"
      name="title"
      onChange={[Function]}
      type="text"
      validationSchema={
        StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "min": true,
            "required": true,
          },
          "_label": "Title",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value="Mug 1"
    />
    <Input
      error=""
      label="Image URL"
      name="imageURL"
      onChange={[Function]}
      type="text"
      validationSchema={
        StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "min": true,
          },
          "_label": "Image URL",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value="#"
    />
    <TextArea
      error=""
      label="Details"
      name="description"
      onChange={[Function]}
      validationSchema={
        StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "max": true,
            "required": true,
          },
          "_label": "Details",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value="product 1 description"
    />
    <Dropdown
      defaultText="Please choose..."
      error=""
      label="Category"
      labelClasses=""
      name="categoryId"
      onChange={[Function]}
      options={
        Array [
          Object {
            "id": "1",
            "name": "category1",
          },
          Object {
            "id": "2",
            "name": "category2",
          },
        ]
      }
      selectClasses="form-control"
      validationSchema={
        NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "required": true,
          },
          "_label": "Category",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value="2"
      wrapperClasses=""
    />
    <Input
      error=""
      label="Price, CURR1"
      name="currentCurrencyPrice"
      onChange={[Function]}
      type="text"
      validationSchema={
        NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "min": true,
            "required": true,
          },
          "_label": "Price",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value={5}
    />
    <Input
      error=""
      label="Dicount, %"
      name="discount"
      onChange={[Function]}
      type="text"
      validationSchema={
        NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "max": true,
            "min": true,
          },
          "_label": "Discount, %",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value={0}
    />
    <Input
      error=""
      label="Producer"
      name="producer"
      onChange={[Function]}
      type="text"
      validationSchema={
        StringSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "min": true,
            "required": true,
          },
          "_label": "Producer",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "string",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value="test producer 1"
    />
    <Input
      error=""
      label="Rate"
      name="rate"
      onChange={[Function]}
      type="text"
      validationSchema={
        NumberSchema {
          "_blacklist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "_conditions": Array [],
          "_deps": Array [],
          "_exclusive": Object {
            "max": true,
            "min": true,
            "required": true,
          },
          "_label": "Rate",
          "_mutate": undefined,
          "_options": Object {
            "abortEarly": true,
            "recursive": true,
          },
          "_type": "number",
          "_typeError": [Function],
          "_whitelist": RefSet {
            "list": Set {},
            "refs": Map {},
          },
          "tests": Array [
            [Function],
            [Function],
            [Function],
          ],
          "transforms": Array [
            [Function],
          ],
        }
      }
      value="5"
    />
    <button
      className="btn btn-secondary w-100"
      disabled={
        Object {
          "imageURL": "Image URL must be at least 5 characters",
        }
      }
      type="submit"
    >
      Save
    </button>
  </form>
</Fragment>
`;
